frontends/mit-open/src/page-components/TabbedCarousel/TabbedCarousel.tsx

```diff
@@ -0,0 +1,142 @@
+import LearningResourceCard from "../LearningResourceCard/LearningResourceCard"
+
+type DataPanelProps<T extends TabConfig["data"] = TabConfig["data"]> = {
+  dataConfig: T
+  children: ({
+    resources,
+    isLoading,
+  }: {
+    resources: LearningResource[]
+    isLoading: boolean
+  }) => React.ReactNode
+}
+
+const ResourcesData: React.FC<DataPanelProps<ResourceDataSource>> = ({
+  dataConfig,
+  children,
+}) => {
+  const { data, isLoading } = useLearningResourcesList(dataConfig.params)
+  return children({ resources: data?.results ?? [], isLoading })
+}
+
+const UpcomingResourcesData: React.FC<DataPanelProps<UpcomingDataSource>> = ({
+  dataConfig,
+  children,
+}) => {
+  const { data, isLoading } = useLearningResourcesUpcoming(dataConfig.params)
+  return children({ resources: data?.results ?? [], isLoading })
+}
+
+const SearchData: React.FC<DataPanelProps<SearchDataSource>> = ({
+  dataConfig,
+  children,
+}) => {
+  const { data, isLoading } = useLearningResourcesSearch(dataConfig.params)
+  return children({ resources: data?.results ?? [], isLoading })
+}
+
+/**
+ * A wrapper to load data based `TabConfig.data`.
+ *
+ * For each `TabConfig.data.type`, a different API endpoint, and hence
+ * react-query hook, is used. Since hooks can't be called conditionally within
+ * a single component, each type of data is handled in a separate component.
+ */
+const DataPanel: React.FC<DataPanelProps> = ({ dataConfig, children }) => {
+  switch (dataConfig.type) {
+    case "resources":
+      return <ResourcesData dataConfig={dataConfig}>{children}</ResourcesData>
+    case "lr_search":
+      return <SearchData dataConfig={dataConfig}>{children}</SearchData>
+    case "resources_upcoming":
+      return (
+        <UpcomingResourcesData dataConfig={dataConfig}>
+          {children}
+        </UpcomingResourcesData>
+      )
+    default:
+      // @ts-expect-error This will always be an error if the switch statement
+      // is exhaustive since dataConfig will have type `never`
+      throw new Error(`Unknown data type: ${dataConfig.type}`)
+  }
+}
+
+const CarouselStyled = styled(Carousel)`
+  .slider-list {
+    /**
+    Prevent shift while loading.
+    This is a bit arbitrary and would be bettr handled by placeholder "skeleton"
+    cards.
+    */
+    min-height: 354px;
+  }
+`
+
+const LearningResourceCardStyled = styled(LearningResourceCard)({
+  height: "100%",
+})
+
+type TabbedCarouselProps = {
+  config: TabConfig[]
+}
+
+/**
+ * A tabbed carousel that fetches resources based on the configuration provided.
+ *  - each TabConfig generates a tab + tabpanel that pulls data from an API based
+ *    on the config
+ *  - data is lazily when the tabpanel first becomes visible
+ */
+const TabbedCarousel: React.FC<TabbedCarouselProps> = ({ config }) => {
+  const [tab, setTab] = React.useState("0")
+
+  return (
+    <TabContext value={tab}>
+      <TabList onChange={(e, newValue) => setTab(newValue)}>
```

Comment:
Wasn't exactly sure where to put this comment as it's more of a suggestion, but here seems appropriate. Something I noticed here is that when you switch tabs, the currently rendered `Carousel` just instantly vanishes, and then the next one takes some time to load. When it does load, it just instantly appears. This all feels a little jarring. I don't know if it's been discussed in any design meetings, but it might be better if we did something here like what the Reorder functionality does in `SortableList`? While moving items, they have `disabled` set so that the whole UI gets slightly greyed out while it's talking to the server. I realize that this is in place to prevent you from sending off conflicting API requests for ordering, but it also provides a nice visual transition when loading data. Something similar here would be appropriate, in my opinion.

Reply:
(If it wasn't clear this is just a suggestion and can be ignored)

Reply:
I think this is a good suggestion. There are a few ways we could handle it...
1. the panels can be pre-loaded. That's more API requests, but they're in parallel, so seems ok.
2. or we could have a skeleton ui or something for placeholder cards
3. or we could try to leave the existing panel in place until the new data has loaded, grayed out a bit like in the sortable list.

I expect we will do one of these things. But I'm inclined to leave it as is for now. I rather expect the carousel design to change a bit. (Arguably (1) and (3) aren't really design issues.)